### PR TITLE
ROCm warp shuffles and reductions

### PR DESCRIPTION
**Pull request by t-vi**
This adds warp shuffle intrinsics to ROCm and enables reductions.

- There was at least one hardcoded 32 threads per warp assumption in `lower_thread_allreduce`.
- I have tentatively hijacked a couple of cuda codegen tests which were useful in verifying the rocm functioning. I'm not quite sure what to do with that. Having rocm tests in test_target_codegen_cuda might not be intuitive, but code duplication is bad, too. The tests helped find the above hardcoded 32 threads.
- This is my first meddling with intrinsics and producing expressions, so bear with me.

It is built on #5726 because I needed that out of the way to work on rocm as a non-nvptx llvm backend.

### CODE REVIEW

**Comment by t-vi** [comment]
@masahi Could I interest you in this?

@tqchen This is the followup to the discussion of #5600

**Comment by tqchen** [comment]
cc @wpan11nv @masahi

**Comment by t-vi** [comment]
Just a quick note: Of the two commits, only the second is new to this PR. If you want me to rebase this on master, let me know.

File: src/target/llvm/intrin_rule_rocm.cc [hunk 1]
```diff
@@ -40,8 +41,59 @@ inline void DispatchExternOCML(const TVMArgs& args, TVMRetValue* rv) {
   *rv = CallNode::make(call->dtype, intrinsic_name.str(), call->args, CallNode::PureExtern);
 }
 
+inline void DispatchShuffle(const TVMArgs& targs, TVMRetValue* rv) {
+  PrimExpr e = targs[0];
```
Comment (wpan11nv) [inline src/target/llvm/intrin_rule_rocm.cc:45]: Let's use more descriptive names for all arguments.

File: tests/python/unittest/test_tir_transform_lower_warp_memory.py [hunk 2]
```diff
@@ -249,9 +245,13 @@ def check(m):
...
             tvm.testing.assert_allclose(B_nd.asnumpy(), B_np)
 
-    check(m=31)
-    check(m=32)
-    check(m=33)
+    for device in ['cuda', 'rocm']:
+        if not tvm.context(device, 0).exist or not tvm.runtime.enabled(device):
+            print("skip because", device,"is not enabled..")
+            continue
+        check(device, m=31)
```
Comment (wpan11nv) [inline tests/python/unittest/test_tir_transform_lower_warp_memory.py:252]: For rocm you may want to check 63,64,65 as it covers three cases comparing to the warp size.

**Comment by tqchen** [comment]
@t-vi yes please rebase to master while addressing the comments.

File: src/target/llvm/intrin_rule_rocm.cc (hunk 1 above)
Comment (t-vi) [reply src/target/llvm/intrin_rule_rocm.cc:45]: While I do agree in general, the e here is more or less a cast and I took this straight from `intrin_rule_llvm.h`, including the naming.

File: tests/python/unittest/test_tir_transform_lower_warp_memory.py (hunk 2 above)
Comment (t-vi) [reply tests/python/unittest/test_tir_transform_lower_warp_memory.py:252]: Tank you for your comments! I added both and rebased.

**Comment by masahi** [comment]
@t-vi This is great! Does this mean rocm backend would use shuffle instruction to compute reduction, such as in softmax?

**Comment by t-vi** [comment]
That's the idea, yes. In my microbenchmark of the imagenet softmax on the Radeon VII, I'm going from ~140µs to ~14µs. The baseline from PyTorch (handcrafted but somewhat generic kernel) is ~18µs, so this is going well. :slightly_smiling_face: 
Of course, the topi work is entirely @wpan11nv 's. I'm quite happy I managed to enable warp reductions on ROCm, though.

File: src/tir/transforms/lower_thread_allreduce.cc [hunk 3]
```diff
@@ -478,9 +478,20 @@ class ThreadAllreduceBuilder final : public StmtExprMutator {
   // the warp size.
   //
   // TODO(tvm-team) reduction with a sub-warp of 8 or 16 threads.
+  // Note: The ROCm backend will only have warp reductions for now.
+  // Also, the warp/wavefron size differs (64 on rocm, 32 on cuda).
```
Comment (masahi) [inline src/tir/transforms/lower_thread_allreduce.cc:482]: wavefront

File: src/tir/transforms/lower_thread_allreduce.cc [hunk 4]
```diff
@@ -478,9 +478,20 @@ class ThreadAllreduceBuilder final : public StmtExprMutator {
...
   bool is_warp_reduction(const std::vector<DataType>& types) const {
     // Only cuda target supports warp reductions.
-    if (target_->target_name != "cuda") return false;
+    if ((target_->target_name != "cuda") && (target_->target_name != "rocm")) return false;
+
+    // rocm only supports 32 bit operands afor shuffeling at the moment
```
Comment (masahi) [inline src/tir/transforms/lower_thread_allreduce.cc:487]: afor -> for
shuffeling -> shuffling

**Comment by masahi** [comment]
Thanks @t-vi @wpan11nv @tqchen
